File: src/api/routes/permanent_absences.js
const express = require('express');
const router = express.Router();
const db = require('../db');

// Get all permanent absences
router.get('/', async (req, res) => {
    try {
        const { rows } = await db.query('SELECT * FROM permanent_absences');
        res.json(rows);
    } catch (err) {
        console.error(err);
        res.status(500).send('Server Error');
    }
});

// Get permanent absences for a student
router.get('/:studentId', async (req, res) => {
    try {
        const { studentId } = req.params;
        const { rows } = await db.query('SELECT * FROM permanent_absences WHERE student_id = $1', [studentId]);
        res.json(rows);
    } catch (err) {
        console.error(err);
        res.status(500).send('Server Error');
    }
});

// Create a new permanent absence
router.post('/', async (req, res) => {
    try {
        const { student_id, weekday, reason } = req.body;
        const { rows } = await db.query(
            'INSERT INTO permanent_absences (student_id, weekday, reason) VALUES ($1, $2, $3) RETURNING *',
            [student_id, weekday, reason]
        );
        res.status(201).json(rows[0]);
    } catch (err) {
        console.error(err);
        res.status(500).send('Server Error');
    }
});

// Update a permanent absence
router.put('/:id', async (req, res) => {
    try {
        const { id } = req.params;
        const { student_id, weekday, reason } = req.body;
        const { rows } = await db.query(
            'UPDATE permanent_absences SET student_id = $1, weekday = $2, reason = $3 WHERE id = $4 RETURNING *',
            [student_id, weekday, reason, id]
        );
        if (rows.length === 0) {
            return res.status(404).send('Permanent absence not found');
        }
        res.json(rows[0]);
    } catch (err) {
        console.error(err);
        res.status(500).send('Server Error');
    }
});

// Delete a permanent absence
router.delete('/:id', async (req, res) => {
    try {
        const { id } = req.params;
        const { rowCount } = await db.query('DELETE FROM permanent_absences WHERE id = $1', [id]);
        if (rowCount === 0) {
            return res.status(404).send('Permanent absence not found');
        }
        res.status(204).send();
    } catch (err) {
        console.error(err);
        res.status(500).send('Server Error');
    }
});

module.exports = router;

File: backend/app/api/permanent_absences.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from ..models.base import get_db
from ..models.permanent_absence import PermanentAbsence
from ..models.student import Student
from ..schemas.permanent_absence import PermanentAbsenceCreate, PermanentAbsenceUpdate, PermanentAbsenceResponse
from ..services.audit_service import log_audit

router = APIRouter()

@router.get("/", response_model=List[PermanentAbsenceResponse])
def get_permanent_absences(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    absences = db.query(PermanentAbsence).offset(skip).limit(limit).all()
    return absences

@router.get("/{absence_id}", response_model=PermanentAbsenceResponse)
def get_permanent_absence(absence_id: int, db: Session = Depends(get_db)):
    absence = db.query(PermanentAbsence).filter(PermanentAbsence.id == absence_id).first()
    if not absence:
        raise HTTPException(status_code=404, detail="Permanent absence not found")
    return absence

@router.post("/", response_model=PermanentAbsenceResponse, status_code=status.HTTP_201_CREATED)
def create_permanent_absence(absence: PermanentAbsenceCreate, db: Session = Depends(get_db)):
    # Check if student exists
    student = db.query(Student).filter(Student.id == absence.student_id).first()
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    
    # Check if permanent absence already exists for this student and weekday
    existing = db.query(PermanentAbsence).filter(
        PermanentAbsence.student_id == absence.student_id,
        PermanentAbsence.weekday == absence.weekday
    ).first()
    if existing:
        raise HTTPException(status_code=400, detail="Permanent absence already exists for this student and weekday")
    
    db_absence = PermanentAbsence(**absence.dict())
    db.add(db_absence)
    db.commit()
    db.refresh(db_absence)
    
    log_audit(db, "manager", "create", "permanent_absence", db_absence.id, None, absence.dict())
    
    return db_absence

@router.put("/{absence_id}", response_model=PermanentAbsenceResponse)
def update_permanent_absence(absence_id: int, absence: PermanentAbsenceUpdate, db: Session = Depends(get_db)):
    db_absence = db.query(PermanentAbsence).filter(PermanentAbsence.id == absence_id).first()
    if not db_absence:
        raise HTTPException(status_code=404, detail="Permanent absence not found")
    
    original_data = {
        "weekday": db_absence.weekday.value if db_absence.weekday else None,
        "reason": db_absence.reason
    }
    
    # Check for unique constraint if weekday is being updated
    if absence.weekday and absence.weekday != db_absence.weekday:
        existing = db.query(PermanentAbsence).filter(
            PermanentAbsence.student_id == db_absence.student_id,
            PermanentAbsence.weekday == absence.weekday
        ).first()
        if existing:
            raise HTTPException(status_code=400, detail="Permanent absence already exists for this student and weekday")
    
    update_data = absence.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_absence, field, value)
    
    db.commit()
    db.refresh(db_absence)
    
    log_audit(db, "manager", "update", "permanent_absence", absence_id, original_data, update_data)
    
    return db_absence

@router.delete("/{absence_id}")
def delete_permanent_absence(absence_id: int, db: Session = Depends(get_db)):
    db_absence = db.query(PermanentAbsence).filter(PermanentAbsence.id == absence_id).first()
    if not db_absence:
        raise HTTPException(status_code=404, detail="Permanent absence not found")
    
    original_data = {
        "student_id": db_absence.student_id,
        "weekday": db_absence.weekday.value if db_absence.weekday else None,
        "reason": db_absence.reason
    }
    
    db.delete(db_absence)
    db.commit()
    
    log_audit(db, "manager", "delete", "permanent_absence", absence_id, original_data, None)
    
    return {"message": "Permanent absence deleted successfully"}

@router.get("/student/{student_id}", response_model=List[PermanentAbsenceResponse])
def get_student_permanent_absences(student_id: int, db: Session = Depends(get_db)):
    absences = db.query(PermanentAbsence).filter(PermanentAbsence.student_id == student_id).all()
    return absences